Capture-recapture-classify using genotypes for individual identity

###4/15/2024 Disclaimer: I realized the categorical ID update wasn't correct, so I switched back to a Metropolis-Hastings ID update like I was using in all other of my latent ID samplers. I will retest with more sparse genotype data and report back. I should have known the categorical sampler wouldn't work, there is no distribution on ID, only y.true. D'oh!


Nonspatial application of the approach used in the genotype SPIM repo, with some differences. No spatial information is used (obviously), observation model is different (Poisson more realistic with space, less so without), and therefore the ID update is different. The observation model here is Bernoulli for detection then zero-truncated Poisson for number of samples given detection. I use a categorical sampler here instead of Metropolis-Hastings I used in genotype SPIM. I like this observation model better and should add it to genotype SPIM. Though that would require not aggregating detections over occasions as I do now (relying on Poisson results).

Currently there are files for M0 and Mh (regular models for microsatellites). For M0, there are versions that do (sampType in file names) and do not allow variation in genotyping error probabilities as a function of categorical covariates. Continuous covariates can be used, but I haven't provided that here. Continuous covariates require a genotype classification matrix for every sample, so it is slow. There are also M0 versions with and without sample type covariates for SNPs.

Note, you need "better data" for Mh as you do for regular capture-recapture. If individual heterogeneity is largely due to spatial juxtoposition of individuals and detectors, and you have the requisite spatial information recorded, perhaps try the genotype SPIM.

Another note: if the observed genotype data is very sparse and the false positive probabilities are initialized near 1, particularly for heterozygotes, it looks like it can get stuck there. The posterior is multimodal, but this mode is obviously the wrong one. In this case, initialize the genotyping error rates to realistic values or change the prior to reflect that false alleles are very rare.